contracts/GroupVoteLocker.sol

```diff
@@ -0,0 +1,331 @@
+        require(_blockNumber <= block.number, "Block number is in the future");
+    
+        // Get most recent global Checkpoint to block
+        uint256 recentGroupStateEpoch = _findEpoch(
+            groupState.checkpoints,
+            _blockNumber,
+            groupState.epoch
+        );
+
+        if (recentGroupStateEpoch == 0) {
+            return 0;
+        }
+
+        Checkpoint memory checkpoint0 = groupState.checkpoints[recentGroupStateEpoch];
+
+        if (checkpoint0.blk > _blockNumber) {
+            return 0;
+        }
+
+        uint256 dTime = 0;
+        if (recentGroupStateEpoch < groupState.epoch) {
+            Checkpoint memory checkpoint1 = groupState.checkpoints[
+                recentGroupStateEpoch + 1
+            ];
+            if (checkpoint0.blk != checkpoint1.blk) {
+                /* to estimate how much time has passed since the last checkpoint get the number
+                 * of blocks since the last checkpoint. And multiply that by the average time per 
+                 * block of the 2 neighboring checkpoints of said _blockNumber
+                 */
+                dTime =
+                    ((_blockNumber - checkpoint0.blk) *
+                        (checkpoint1.ts - checkpoint0.ts)) /
+                    (checkpoint1.blk - checkpoint0.blk);
+            }
+        } else if (checkpoint0.blk != block.number) {
+            /* to estimate how much time has passed since the last checkpoint get the number
+             * of blocks since the last checkpoint. And multiply that by the average time per 
+             * block since the last checkpoint and present blockchain state. 
+             */
+            dTime =
+                ((_blockNumber - checkpoint0.blk) *
+                    (block.timestamp - checkpoint0.ts)) /
+                (block.number - checkpoint0.blk);
+        }
+        // if code doesn't enter any of the above if conditions latest _blockNumber was passed
+        // to the function and dTime is correctly set to 0
+
+        // Now dTime contains info on how far are we beyond point
+        return _votePowerAt(groupState, checkpoint0, checkpoint0.ts + dTime);
+    }
+
+    /**
+     * @dev Calculates voting power at a given time _t
+     * @param groupState group state to calculate voting power from
+     * @param _checkpoint Most recent point before time _t
+     * @param _time Time at which to calculate supply
+     * @return totalSupply at given time
+     */
+    function _votePowerAt(GroupVotePowerState storage groupState, Checkpoint memory _checkpoint, uint256 _time)
+        internal
+        view
+        returns (uint256)
+    {
+        Checkpoint memory lastCheckpoint = _checkpoint;
+
+        // Floor the timestamp to weekly interval
+        uint256 iterativeTime = _floorToWeek(lastCheckpoint.ts);
+        // Iterate through all weeks between _checkpoint & _time to account for slope changes
+        for (uint256 i = 0; i < 255; i++) {
```

Comment:
So this will break everything if we go for five years without a checkpoint? 

Do have we add a way to force adding intermediate checkpoints from some method?

Reply:
Can we skip this loop if instead of a mapping to group slope changes, we somehow work of the checkpoints that have both a bias and a slope?